arq/worker.py

```diff
@@ -757,8 +764,30 @@ def _add_signal_handler(self, signum: Signals, handler: Callable[[Signals], None
     def _jobs_started(self) -> int:
         return self.jobs_complete + self.jobs_retried + self.jobs_failed + len(self.tasks)
 
+    async def _wait_for_tasks_to_complete(self) -> None:
+        while self.tasks:
+            await asyncio.sleep(0.1)
+
     def handle_sig(self, signum: Signals) -> None:
         sig = Signals(signum)
+        if self._wait_for_job_completion_on_signal_seconds:
+            logger.info('Setting allow_pick_jobs to `False`')
+            self.allow_pick_jobs = False
+            logger.info(
+                'shutdown on %s ◆ %d jobs complete ◆ %d failed ◆ %d retries ◆ %d ongoing waiting for completion',
+                sig.name,
+                self.jobs_complete,
+                self.jobs_failed,
+                self.jobs_retried,
+                len(self.tasks),
+            )
+            with contextlib.suppress(TimeoutError):
+                self.loop.create_task(
+                    asyncio.wait_for(
+                        self._wait_for_tasks_to_complete(),
+                        self._wait_for_job_completion_on_signal_seconds,
+                    )
+                )
```

Comment:
Converting to draft again, I see that this don't actually work as intended. This creates a task, but don't wait for it to complete.

```python
            with contextlib.suppress(TimeoutError):
                logger.info('Waiting %d seconds for tasks to complete', self._wait_for_job_completion_on_signal_seconds)
                wait_for_task = self.loop.create_task(
                    asyncio.wait_for(
                        self._wait_for_tasks_to_complete(),
                        self._wait_for_job_completion_on_signal_seconds,
                    )
                )
                self.loop.run_until_complete(wait_for_task)
```
also don't work. 

I'll have to rewrite to be more like the previous commit I guess, but I don't think I'll make it for this release.

Reply:
Understood, thanks for letting me know.

Reply:
Actually, I think I've made it. Give me one sec. I'll fix other feedback too.